hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClusterConnectorServiceImpl.java

```diff
@@ -361,18 +361,16 @@ public void connectionAdded(Connection connection) {
     @Override
     public void connectionRemoved(Connection connection) {
         final ClientConnection clientConnection = (ClientConnection) connection;
-        if (clientConnection.isAuthenticatedAsOwner()) {
+        // if cluster listening connection is disconnected
+        if (clientConnection == clusterListeningConnection) {
             clusterConnectionExecutor.execute(new Runnable() {
                 @Override
                 public void run() {
-                    Address endpoint = clientConnection.getEndPoint();
-                    // it may be possible that while waiting on executor queue, the client got connected (another connection),
-                    // then we do not need to do anything for cluster disconnect.
-                    if (endpoint == null || !endpoint.equals(ownerConnectionAddress)) {
+                    if (clientConnection != clusterListeningConnection) {
```

Comment:
Also check if the number of activeConnections is zero. If there is at least one other connection, then do not disconnect from cluster but just register the partition and membership listeners on that other alive connection and do not change state but stay as CLIENT_CONNECTED.